hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/KeyGenUtils.java

```diff
@@ -67,28 +67,33 @@ public static KeyGeneratorType inferKeyGeneratorType(
       Option<String> recordsKeyFields, String partitionFields) {
     boolean autoGenerateRecordKeys = !recordsKeyFields.isPresent();
     if (autoGenerateRecordKeys) {
-      return inferKeyGeneratorTypeForAutoKeyGen(partitionFields);
+      return inferKeyGeneratorTypeFromPartitionFields(partitionFields);
     } else {
       if (!StringUtils.isNullOrEmpty(partitionFields)) {
-        int numPartFields = partitionFields.split(",").length;
         int numRecordKeyFields = recordsKeyFields.get().split(",").length;
-        if (numPartFields == 1 && numRecordKeyFields == 1) {
-          return KeyGeneratorType.SIMPLE;
+        KeyGeneratorType recordKeyGeneratorType = numRecordKeyFields == 1 ? KeyGeneratorType.SIMPLE : KeyGeneratorType.COMPLEX;
+        KeyGeneratorType partitionKeyGeneratorType = inferKeyGeneratorTypeFromPartitionFields(partitionFields);
+        if (partitionKeyGeneratorType == KeyGeneratorType.SIMPLE && recordKeyGeneratorType == KeyGeneratorType.COMPLEX) {
+          return recordKeyGeneratorType;
+        } else {
+          return partitionKeyGeneratorType;
         }
-        return KeyGeneratorType.COMPLEX;
       }
       return KeyGeneratorType.NON_PARTITION;
     }
   }
 
   // When auto record key gen is enabled, our inference will be based on partition path only.
-  private static KeyGeneratorType inferKeyGeneratorTypeForAutoKeyGen(String partitionFields) {
+  private static KeyGeneratorType inferKeyGeneratorTypeFromPartitionFields(String partitionFields) {
```

Comment:
Let's add unit tests for the static methods in this class. Tests should cover all different keygen types.

Reply:
Addressed